crates/ruff_linter/src/rules/ruff/rules/suppression_comment_visitor.rs

```diff
@@ -0,0 +1,313 @@
+use std::iter::{FilterMap, Peekable};
+
+use ruff_python_ast::{
+    visitor::preorder::{self, PreorderVisitor, TraversalSignal},
+    AnyNodeRef, Suite,
+};
+use ruff_python_trivia::{
+    indentation_at_offset, CommentLinePosition, CommentRanges, SimpleTokenKind, SimpleTokenizer,
+    SuppressionKind,
+};
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextLen, TextRange, TextSize};
+
+#[derive(Clone, Copy, Debug)]
+struct SuppressionComment {
+    range: TextRange,
+    kind: SuppressionKind,
+}
+
+type MapCommentFn<'src> = Box<dyn Fn(&'src TextRange) -> Option<SuppressionComment> + 'src>;
+type CommentIter<'src> = Peekable<FilterMap<std::slice::Iter<'src, TextRange>, MapCommentFn<'src>>>;
+
+/// Visitor that captures AST data for suppression comments. This uses a similar approach
+/// to `CommentsVisitor` in the formatter crate.
+pub(super) struct SuppressionCommentVisitor<'src, 'builder> {
+    comments: CommentIter<'src>,
+
+    parents: Vec<AnyNodeRef<'src>>,
+    preceding_node: Option<AnyNodeRef<'src>>,
+    comments_in_scope: Vec<(Option<AnyNodeRef<'src>>, SuppressionKind)>,
+
```

Comment:
It seems we always push `Some(node)`, so using `Option<AnyNodeRef>` here is unnecessary.